clap_derive/src/derives/args.rs

```diff
@@ -51,10 +52,188 @@ pub(crate) fn derive_args(input: &DeriveInput) -> Result<TokenStream, syn::Error
                 .collect::<Result<Vec<_>, syn::Error>>()?;
             gen_for_struct(&item, ident, &input.generics, &fields)
         }
+        Data::Enum(DataEnum { ref variants, .. }) => {
+            let name = Name::Derived(ident.clone());
+            let item = Item::from_args_struct(input, name)?;
+
+            let variant_items = variants
+                .iter()
+                .map(|variant| {
+                    let item =
+                        Item::from_args_enum_variant(variant, item.casing(), item.env_casing())?;
+                    Ok((item, variant))
+                })
+                .collect::<Result<Vec<_>, syn::Error>>()?;
+
+            gen_for_enum(&item, ident, &input.generics, &variant_items)
+        }
         _ => abort_call_site!("`#[derive(Args)]` only supports non-tuple structs"),
     }
 }
 
+pub(crate) fn gen_for_enum(
+    _item: &Item,
+    item_name: &Ident,
+    generics: &Generics,
+    variants: &[(Item, &Variant)],
+) -> Result<TokenStream, syn::Error> {
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let app_var = Ident::new("__clap_app", Span::call_site());
+    let mut augmentations = TokenStream::default();
+    let mut augmentations_update = TokenStream::default();
+
+    let mut constructors = TokenStream::default();
+    let mut updaters = TokenStream::default();
+
+    for (item, variant) in variants.iter() {
+        let Fields::Named(ref fields) = variant.fields else {
+            abort! { variant.span(),
+                "`#[derive(Args)]` only supports named enum variants if used on an enum",
+            }
+        };
+        let group_id = item.group_id();
+
+        let conflicts = variants
+            .iter()
+            .filter_map(|(_, v)| {
+                if v.ident == variant.ident {
+                    None
+                } else {
+                    Some(Name::Derived(v.ident.clone()))
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let fields = collect_args_fields(item, fields)?;
+
+        let augmentation = gen_augment(&fields, &app_var, item, &conflicts, false)?;
```

Comment:
Ah specify multiple(false) for the "outer" group generated for the enum itself?
Guess that could work.
I explicitly stated the conflicts more out of intuition and reading that nested groups are behaving weirdly.

Reply:
So to elaborate on "behaving weirdly", adding a group as an argument of a group doesnt seem to work at all currently

if i build something like this:
```
Command::new()
	.arg(clap::Arg::new("a") ...)
	.arg(clap::Arg::new("b") ...)
	.group(clap::ArgGroup("A").args(["a"])
	.group(clap::ArgGroup("B").args(["b"])
	.group(clap::ArgGroup("Outer").args(["A", "B"])
```

i get a runtime error: `Command clap: Argument group 'Outer' contains non-existent argument 'A'`.
Before I'm digging into the ArgGroup parser which seems to specifically address _args_ (rather than _args or groups_), i'm looking to confirm that this is indeed how we would want to specify nested groups to begin with.

-----------

Continue reading here if i haven't gone astray yet.

From thereon i'm looking at `clap_builder::parser::validator::gather_arg_direct_conflicts` which would recursively resolve all ancestor chains, and then from the root(s) down check each group whether it is a `multiple == false` group, and if it is add all args from other branches not on the current path as conflicts; or in pseudocode:


```
conflicts(arg_id: &Id, path: Vec<&Id>, conf: &mut Vec<Id>)

parent_id = path[-1] ? arg_id
for group in groups_for_arg(parnt_id) {
	
	for conflict in group.conflicts {
		if conflict.is_group {
			for member in members_recursive_ignore_path(path, group) {
				conf.push(member.id)
			}
		} else {
			conf.push(conflict.id)
		}
	}

	if !group.multiple {
		for member in members_recursive_ignore_path(path, group) {
			conf.push(member.id)
		}
	}

	path = path.clone()
	path.push(group.id)
	conflicts(arg_id, path, conf)
}
```

that is if the goal is to support arbitrary nesting
I'm not too happy with the amount of potential duplicates iff groups are densely nested.
Also keep calling `groups_for_arg` (or the impl of it) may hold potential for optimization down the road.

I believe a similar impl is also necessary for `require`ing with nested groups.

Reply:
Oh wow, I thought we supported that.  There was some case where we had a bug related to `ArgGroup`s and I thought we had broader support for `ArgGroup`s than we apparently do (which also means I blocked #4697 on the wrong thing...)

I would like for this to move forward with groups, rather than conflicts, so we make sure we stabilize the right semantics.  So that means we need nested groups first...

One approach
- Deprecate `ArgGroup::arg` and `ArgGroup::args` and add `ArgGroup::member` or `ArgGroup::child`
- Update the assertions related to ensuring group members are present
- Add associated tests to make sure nested arg groups work, fixing bugs along the way

If someone wants to take this on, this should be its own PR.

Reply:
I'd be up to look into that.
Do you think my understanding of conflict detection aligns with your intended changes to arggroup?
How/where could we plan this in more detail then 3 bullet points?

Reply:
Created #5711.  I don't have more specific guidance at this time.  In general, a child group should behave like a child arg.  I don't more more specific guidance than that at this time.